### PR TITLE
Remove `abs` from `DefaultQubit`

### LINKED ISSUE #1125: Incorrect value of gradients at 0 after taking gradient twice
#### Issue description

A colleague noticed that when running `qml.grad` twice on a particular circuit, to compute a second derivative, the output value of the gradient is incorrect when the input value is precisely 0. The specific circuit is effectively the function `cos(x)` (see below for the code to reproduce).

* *Expected behavior:* The value of the second gradient at 0 for this function should be -1. 

* *Actual behavior:* The value of the second gradient at 0 is computed to be -0.5.

* *Reproduces how often:* Always 

* *System information:* (post the output of `import pennylane as qml; qml.about()`)

```
Name: PennyLane
Version: 0.15.0.dev0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /home/olivia/Software/anaconda3/envs/xanadu/lib/python3.8/site-packages
Requires: appdirs, networkx, scipy, autograd, toml, semantic-version, numpy
Required-by: PennyLane-SF, pennylane-qulacs, PennyLane-qsharp, PennyLane-qiskit, PennyLane-Qchem, PennyLane-Forest, PennyLane-Cirq
Platform info:           Linux-5.4.0-66-generic-x86_64-with-glibc2.10
Python version:          3.8.5
Numpy version:           1.18.5
Scipy version:           1.4.1
Installed devices:
- default.gaussian (PennyLane-0.15.0.dev0)
- default.mixed (PennyLane-0.15.0.dev0)
- default.qubit (PennyLane-0.15.0.dev0)
- default.qubit.autograd (PennyLane-0.15.0.dev0)
- default.qubit.jax (PennyLane-0.15.0.dev0)
- default.qubit.tf (PennyLane-0.15.0.dev0)
- default.tensor (PennyLane-0.15.0.dev0)
- default.tensor.tf (PennyLane-0.15.0.dev0)
- strawberryfields.fock (PennyLane-SF-0.13.0.dev0)
- strawberryfields.gaussian (PennyLane-SF-0.13.0.dev0)
- strawberryfields.gbs (PennyLane-SF-0.13.0.dev0)
- strawberryfields.remote (PennyLane-SF-0.13.0.dev0)
- strawberryfields.tf (PennyLane-SF-0.13.0.dev0)
- qulacs.simulator (pennylane-qulacs-0.12.0)
- microsoft.QuantumSimulator (PennyLane-qsharp-0.8.0)
- qiskit.aer (PennyLane-qiskit-0.14.0.dev0)
- qiskit.basicaer (PennyLane-qiskit-0.14.0.dev0)
- qiskit.ibmq (PennyLane-qiskit-0.14.0.dev0)
- forest.numpy_wavefunction (PennyLane-Forest-0.13.0.dev0)
- forest.qpu (PennyLane-Forest-0.13.0.dev0)
- forest.qvm (PennyLane-Forest-0.13.0.dev0)
- forest.wavefunction (PennyLane-Forest-0.13.0.dev0)
- cirq.mixedsimulator (PennyLane-Cirq-0.13.0)
- cirq.pasqal (PennyLane-Cirq-0.13.0)
- cirq.qsim (PennyLane-Cirq-0.13.0)
- cirq.qsimh (PennyLane-Cirq-0.13.0)
- cirq.simulator (PennyLane-Cirq-0.13.0)
```

#### Source code and tracebacks

Code to reproduce:

```
import pennylane as qml
from pennylane import numpy as np

dev = qml.device("default.qubit", wires=1)

# Produces function cos(x)
@qml.qnode(dev)
def circuit(x):
    qml.RY(x, wires=0)
    return qml.expval(qml.PauliZ(0))

gradcos  = qml.grad(circuit)  # -sin(x)
gradcos2 = qml.grad(gradcos)  # -cos(x)

print(gradcos2(0.)) # Should be -1, but outputs -0.5

```

The issue seems to be at 0 specifically, as evaluating the gradient at `1e-6` yields a result very close to -1. 

I tried another example as well:
```
# Produces function -i*sin(x)
@qml.qnode(dev)
def circuit(x):
    qml.PauliX(wires=0)
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))

gradcos  = qml.grad(circuit)  # -i * cos(x)
gradcos2 = qml.grad(gradcos)  # i * sin(x)

print(gradcos2(0.)) # Should be 1, but outputs 0.5
```

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
This PR implements the fix to #1125 proposed in [this comment](https://github.com/PennyLaneAI/pennylane/issues/1125#issuecomment-999206195)

**Description of the Change:**
The line 792
```python
    prob = self.marginal_prob(self._abs(self._flatten(self._state)) ** 2, wires)
```
in `qml/devices/default_qubit.py` is replaced by
```python
    flat_state = self._flatten(self._state)
    real_state = self._real(flat_state)
    imag_state = self._imag(flat_state)
    prob = self.marginal_prob(real_state ** 2 + imag_state ** 2), wires)
```
avoiding the usage of `abs`, which does not have a well-defined derivative at 0.

**Benefits:**
Fixes #1125 in all interfaces.

**Possible Drawbacks:**
Performance, see short discussion [here](https://github.com/PennyLaneAI/pennylane/issues/1125#issuecomment-999206195)

**Related GitHub Issues:**
#1125

### CODE REVIEW

File: pennylane/devices/default_qubit.py [hunk 1]
```diff
@@ -789,5 +789,8 @@ def analytic_probability(self, wires=None):
         if self._state is None:
             return None
 
-        prob = self.marginal_prob(self._abs(self._flatten(self._state)) ** 2, wires)
+        flat_state = self._flatten(self._state)
+        real_state = self._real(flat_state)
+        imag_state = self._imag(flat_state)
+        prob = self.marginal_prob(real_state ** 2 + imag_state ** 2, wires)
```
Comment (josh146) [inline pennylane/devices/default_qubit.py:795]: 🙌

File: pennylane/devices/default_qubit_autograd.py [hunk 2]
```diff
@@ -92,6 +92,7 @@ class DefaultQubitAutograd(DefaultQubit):
     _transpose = staticmethod(np.transpose)
     _tensordot = staticmethod(np.tensordot)
     _conj = staticmethod(np.conj)
+    _real = staticmethod(np.real)
```
Comment (josh146) [inline pennylane/devices/default_qubit_autograd.py:95]: Minor, but we are in the process of slowly migrating the devices to `qml.math` :) Not an issue here though!

File: tests/devices/test_default_qubit_autograd.py [hunk 3]
```diff
@@ -280,6 +280,22 @@ def circuit(a, b):
         )
         assert np.allclose(res, expected_grad, atol=tol, rtol=0)
 
+    @pytest.mark.parametrize("x, shift", [(0.0, 0.0), (0.5, -0.5)])
+    def test_hessian_at_zero(self, x, shift):
+        """Tests that the Hessian at vanishing state vector amplitudes
+        is correct."""
+        dev = qml.device("default.qubit.autograd", wires=1)
+
+        @qml.qnode(dev, interface="autograd")
```
Comment (josh146) [inline tests/devices/test_default_qubit_autograd.py:289]:
```suggestion
        @qml.qnode(dev, interface="autograd", diff_method="backprop")
```
(to protect this test in case defaults change)

File: tests/devices/test_default_qubit_jax.py [hunk 4]
```diff
@@ -543,6 +543,22 @@ def circuit(a, b):
 
         assert jnp.allclose(jnp.array(res), jnp.array(expected_grad), atol=tol, rtol=0)
 
+    @pytest.mark.parametrize("x, shift", [(0.0, 0.0), (0.5, -0.5)])
+    def test_hessian_at_zero(self, x, shift):
+        """Tests that the Hessian at vanishing state vector amplitudes
+        is correct."""
+        dev = qml.device("default.qubit.jax", wires=1)
+
+        @qml.qnode(dev, interface="jax")
```
Comment (josh146) [inline tests/devices/test_default_qubit_jax.py:552]:
```suggestion
        @qml.qnode(dev, interface="jax", diff_method="backprop")
```

File: tests/devices/test_default_qubit_tf.py [hunk 5]
```diff
@@ -1323,6 +1323,33 @@ def circuit(a, b):
...
+    @pytest.mark.parametrize("x, shift", [(0.0, 0.0), (0.5, -0.5)])
+    def test_hessian_at_zero(self, x, shift):
+        """Tests that the Hessian at vanishing state vector amplitudes
+        is correct."""
+        dev = qml.device("default.qubit.tf", wires=1)
+
+        shift = tf.constant(shift)
+        x = tf.Variable(x)
+
+        @qml.qnode(dev, interface="tf")
```
Comment (josh146) [inline tests/devices/test_default_qubit_tf.py:1335]:
```suggestion
        @qml.qnode(dev, interface="tf", diff_method="backprop")
```

File: tests/devices/test_default_qubit_torch.py [hunk 6]
```diff
@@ -1425,6 +1425,26 @@ def circuit(a, b):
...
 
+    @pytest.mark.parametrize("x, shift", [(0.0, 0.0), (0.5, -0.5)])
+    def test_hessian_at_zero(self, torch_device, x, shift):
+        """Tests that the Hessian at vanishing state vector amplitudes
+        is correct."""
+        dev = qml.device("default.qubit.torch", wires=1, torch_device=torch_device)
+
+        x = torch.tensor(x, requires_grad=True)
+
+        @qml.qnode(dev, interface="torch")
```
Comment (josh146) [inline tests/devices/test_default_qubit_torch.py:1436]:
```suggestion
        @qml.qnode(dev, interface="torch", diff_method="backprop")
```

**Review by josh146 — APPROVED** [review]
That was super quick, thanks @dwierichs! I agree, this is good to have fixed even with the minor overhead.

**Comment by dwierichs** [comment]
Thanks for the fast review! And very good idea to make the tests stable against possible default changes. :)
